src/reducers/index.js

```diff
@@ -1,12 +1,63 @@
 import { enableBatching } from 'utils/redux';
-import systemReducer from 'reducers/system';
-import feedsReducer from 'reducers/feeds';
-import cdpsReducer from 'reducers/cdps';
-
-const rootReducer = ({ system, feeds, cdps }, action) => ({
-  system: systemReducer(system, action),
-  feeds: feedsReducer(feeds, action),
-  cdps: cdpsReducer(cdps, action)
-});
+import produce from 'immer';
+import setWith from 'lodash/setWith';
+import * as math from '@makerdao/dai-plugin-mcd/dist/math';
+import ilkList from '../references/ilkList';
+import { SPOT, LIQUIDATION_RATIO, FEED_VALUE_USD } from './feeds';
+import each from 'lodash/each';
+
+const initialState = { system: {}, urns: {}, ilks: {} };
+
+const rootReducer = produce((draft, action) => {
+  if (action.type === 'CLEAR_CONTRACT_STATE') return initialState;
+
+  if (action.type === 'watcherUpdates') {
+    action.payload.forEach(({ type, value }) => {
+      // FIXME track down why this is happening
+      if (type === 'undefined') return;
+
+      // lodash.setWith creates nested objects from dotted paths
+      setWith(
+        draft,
+        type.replace(/^urn/, 'urns').replace(/^ilk/, 'ilks'),
+        value,
+        Object
+      );
+
+      if (type.startsWith('ilk')) {
+        const [, name, prop] = type.split('.');
+        const ilk = draft.ilks[name];
+
+        if (prop === 'duty') {
+          ilk.stabilityFee = math.annualStabilityFee(value);
+        }
+
+        if ([SPOT, LIQUIDATION_RATIO].includes(prop))
```

Comment:
this is an example of caching a potentially expensive value in the store by manually tracking when its dependencies change